content/blog/2022-04-19-tracing/index.md

```diff
@@ -0,0 +1,142 @@
+
+A graph depicting this approach is shown below:
+
+<p align="center">
+<img src="nested_tree.png" alt="nested Tree Formation" title="nested Tree Formation" style="zoom:25%;" />
+</p>
+
+# TraceMonkey VM Implementation & Optimization
+
+Finally, we take a look at how TraceMonkey's VM implemented. TraceMonkey's VM stores all compiled traces inside trace cache, indexed by interpreter PC and type signatures. The trace monitor builds a activation record when executing a trace, with associated local and global variables, temporary stack spaces, and space for arguments to native calls. The trace could then be called like a normal function pointer. The trace monitor is also responsible for restoring the interpreter state, verifying states for compiled trace and releasing the occupied space.
+
+The author also introduces an intresting optimization technique called Tree Stiching. The purpose of tree stitching is to reuse activation record when a side exit is met. The reason is fairly straightforward, since it's easy to see that within a loop, the register allocation and memory layouts are often invariant. the exiting trace only needs to write live
+register-carried values back to its trace activation record. However, with multi-code processors being the majority (unlike when this paper just got published), it's probably just easier to recompile the entire trace-tree in backend.
+
+Additional optimization techniques when a LIR trace gets compiled to machine code as a set of filters, including
+1. Float to Integer Conversion
+2. Constant Subexpression Elimination
+3. Constant Folding
+4. Dead Code Elimination
+5. Dead Stack Elimination
+
+# Evaluation
+The evaluation of is paper is performed on a Macbook Pro over SunSpider, which is a standard testing suite for JavaScript with a variety of applications. The experimental environment is very reasonable to me, since it's a very good representation of JavaScript workloads and user's device. The author benchmarks TraceMonkey against SpiderMonkey (baseline interpreter), SFX (inline threading interpreter in Apple's Webkit), V8 (Method-based JIT developed by Google).
+
+Based on the figure above, traceMonkey is the fastest VM on 9 out of 26 benchmarks. In particular, TraceMonkey performs on an average of ~5x compared to SFX and V8 on `bitops` benchmarks, which are mainly short bitwise operations. In addition to that, TraceMonkey performs well in maths and crypto, slightly edging SFX and V8. TraceMonkey does not achieve the same level of speedup against SFX and V8 on programs with a high number of traces or tiny loop bodies, as well as cases where TraceMonkey currently not tracing (i.e. recursion), although I believe it's relatively easy and reasonable to implement tracing for recursion.
```

Comment:
> although I believe it's relatively easy and reasonable to implement tracing for recursion.

I think it might be nice to provide a little more detail here about why you think so. Seems like an interesting hypothesis!